src/test/ui/lint/uninitialized-zeroed.rs

```diff
@@ -118,9 +125,14 @@ fn main() {
         let _val: bool = MaybeUninit::zeroed().assume_init();
         let _val: [bool; 0] = MaybeUninit::uninit().assume_init();
         let _val: [!; 0] = MaybeUninit::zeroed().assume_init();
-        // Some things that happen to work due to rustc implementation details,
+
+        // Some things that happen to be UB-free due to rustc implementation details,
         // but are not guaranteed to keep working.
         let _val: i32 = mem::uninitialized();
         let _val: OneFruit = mem::uninitialized();
+
+        // Some things that should be caught, but currently are not.
+        let _val: OneFruitNonzero = mem::zeroed();
+        let _val: OneFruitNonzero = mem::uninitialized();
```

Comment:
Would be nice to improve the lint to detect these as well... @5225225 any good ideas for how to do that? I think for non-`is_multi_variant` enums, we probably can compute the variant that *does* exist, and then we can recursively check if that one is valid? We'd also need a smarter `is_multi_variant` that can handle some dataful variants (ensuring they are definitely inhabited).

Reply:
I had a look at making the lint better a while back, and didn't get very far (though, granted, I didn't look too hard). I'll have a look again tomorrow, we definitely should be able to detect this.

Reply:
Moved this into a separate issue: https://github.com/rust-lang/rust/issues/102043